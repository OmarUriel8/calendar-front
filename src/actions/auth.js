import { fetchConToken, fetchSinToken } from '../helpers/fetch';
import { mostrarErrores } from '../helpers/mostrarErrores';
import { types } from '../types/types';

export const startLogin = (email, password) => {
	return async (dispatch) => {
		const res = await fetchSinToken('auth', { email, password }, 'POST');
		const body = await res.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			localStorage.setItem('token-init-date', new Date().getTime());

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				})
			);
		} else {
			mostrarErrores(body);
		}
	};
};

export const startRegister = (email, name, password) => {
	return async (dispatch) => {
		const res = await fetchSinToken('auth/new', { name, email, password }, 'POST');

		const body = await res.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			localStorage.setItem('token-init-date', new Date().getTime());

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				})
			);
		} else {
			mostrarErrores(body);
		}
	};
};

export const startChecking = () => {
	return async (dispatch) => {
		const res = await fetchConToken('auth/renew');

		const body = await res.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			localStorage.setItem('token-init-date', new Date().getTime());

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				})
			);
		} else {
			// mostrarErrores(body);
			dispatch(checkingFinish());
		}
	};
};

const checkingFinish = () => ({
	type: types.authCheckingFinish,
});

const login = (user) => ({
	type: types.authLogin,
	payload: user,
});

export const startLogout = () => {
	return (dispatch) => {
		localStorage.removeItem('token');
		localStorage.removeItem('token-init-date');

		dispatch(logout());
	};
};

const logout = () => ({
	type: types.authLogout,
});
