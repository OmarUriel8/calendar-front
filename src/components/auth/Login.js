import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startLogin } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './login-register.css';

export const Login = () => {
	const [emailValid, setEmailValid] = useState(true);
	const [passwordValid, setPasswordValid] = useState(true);
	const [loading, setLoading] = useState(false);

	const dispatch = useDispatch();

	const [formValues, handleInputChange] = useForm({
		email: 'correo@correo.com',
		password: '123456',
	});

	const { email, password } = formValues;

	const handleLogin = (e) => {
		let valid = true;
		e.preventDefault();
		setEmailValid(true);
		setPasswordValid(true);
		if (!validator.isEmail(email)) {
			setEmailValid(false);
			valid = false;
		}
		if (!validator.isLength(password, { min: 6 })) {
			setPasswordValid(false);
			valid = false;
		}

		if (valid) {
			setLoading(true);
			dispatch(startLogin(email, password));
		}
	};
	return (
		<div className="container login-container mb-5">
			<div className="row">
				<div className="col-md-6 login-form-1 offset-md-3">
					<h3>Ingreso</h3>
					<form onSubmit={handleLogin}>
						<div className="form-group">
							<input
								type="email"
								className={`form-control ${!emailValid && 'is-invalid'} `}
								placeholder="Correo"
								name="email"
								value={email}
								onChange={handleInputChange}
							/>
						</div>
						<div className="form-group">
							<input
								type="password"
								className={`form-control ${!passwordValid && 'is-invalid'} `}
								placeholder="Contraseña"
								name="password"
								value={password}
								onChange={handleInputChange}
							/>
						</div>
						<div className="form-group">
							<input
								type="submit"
								className={`btnSubmit`}
								value="Login"
								disabled={loading}
							/>
						</div>
						<Link to="/auth/register">¿Aún no tienes una cuenta?, Registrate</Link>
					</form>
				</div>
			</div>
		</div>
	);
};
