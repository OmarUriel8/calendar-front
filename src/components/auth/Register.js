import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import Swal from 'sweetalert2';
import validator from 'validator';
import { startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './login-register.css';

export const Register = () => {
	const [nameValid, setNameValid] = useState(true);
	const [emailValid, setEmailValid] = useState(true);
	const [passwordValid, setPasswordValid] = useState(true);
	const [passwordValid2, setPasswordValid2] = useState(true);
	const [loading, setLoading] = useState(false);

	const dispatch = useDispatch();

	const [formValues, handleInputChange] = useForm({
		name: 'kenny',
		email: 'correo2@correo.com',
		password: '123456',
		password2: '123456',
	});

	const { email, password, password2, name } = formValues;

	const handleRegister = (e) => {
		let valid = true;
		e.preventDefault();
		setEmailValid(true);
		setNameValid(true);
		setPasswordValid(true);
		setPasswordValid2(true);

		if (!validator.isEmail(email)) {
			setEmailValid(false);
			valid = false;
		}

		if (!validator.isAlpha(name)) {
			setNameValid(false);
			valid = false;
		}

		if (!validator.isLength(password, { min: 6 })) {
			setPasswordValid(false);
			valid = false;
		}

		if (!validator.isLength(password2, { min: 6 })) {
			setPasswordValid2(false);
			valid = false;
		}

		if (password !== password2) {
			setPasswordValid2(false);
			Swal.fire('Error', 'Las contraseñas no son iguales', 'error');
			valid = false;
		}

		if (valid) {
			setLoading(true);
			dispatch(startRegister(email, name, password));
		}
	};
	return (
		<div className="container login-container mb-5">
			<div className="row">
				<div className="col-md-6 login-form-2 offset-md-3">
					<h3>Registro</h3>
					<form onSubmit={handleRegister}>
						<div className="form-group">
							<input
								type="text"
								className={`form-control ${!nameValid && 'is-invalid'} `}
								placeholder="Nombre"
								onChange={handleInputChange}
								value={name}
								name="name"
							/>
						</div>
						<div className="form-group">
							<input
								type="email"
								className={`form-control ${!emailValid && 'is-invalid'} `}
								placeholder="Correo"
								onChange={handleInputChange}
								value={email}
								name="email"
							/>
						</div>
						<div className="form-group">
							<input
								type="password"
								className={`form-control ${!passwordValid && 'is-invalid'} `}
								placeholder="Contraseña"
								onChange={handleInputChange}
								value={password}
								name="password"
							/>
						</div>

						<div className="form-group">
							<input
								type="password"
								className={`form-control ${!passwordValid2 && 'is-invalid'} `}
								placeholder="Repita la contraseña"
								onChange={handleInputChange}
								value={password2}
								name="password2"
							/>
						</div>

						<div className="form-group">
							<input
								type="submit"
								className="btnSubmit"
								value="Crear cuenta"
								disabled={loading}
							/>
						</div>
						<Link to="/auth/login" className="back-login">
							¿Ya tienes una cuenta?, Iniciar Sesión
						</Link>
					</form>
				</div>
			</div>
		</div>
	);
};
