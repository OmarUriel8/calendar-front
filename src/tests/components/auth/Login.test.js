import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { Login } from '../../../components/auth/Login';
import { MemoryRouter } from 'react-router';
import { startLogin } from '../../../actions/auth';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
let store = mockStore(initState);
store.dispatch = jest.fn();

jest.mock('../../../actions/auth', () => ({
	startLogin: jest.fn(),
}));

describe('Pruebas en <Login/>', () => {
	const wrapper = mount(
		<Provider store={store}>
			<MemoryRouter>
				<Login />
			</MemoryRouter>
		</Provider>
	);
	beforeEach(() => {
		store = mockStore(initState);
	});

	test('Debe mostrarse correctamente', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('Debe de llamar el dispatch del login', () => {
		wrapper.find('input[name="email"]').simulate('change', {
			target: {
				name: 'email',
				value: 'omar@correo.com',
			},
		});

		wrapper.find('input[name="password"]').simulate('change', {
			target: {
				name: 'password',
				value: 'asdfgh',
			},
		});

		wrapper.find('form').prop('onSubmit')({
			preventDefault: jest.fn(),
		});

		expect(startLogin).toHaveBeenCalledWith('omar@correo.com', 'asdfgh');
	});
});
