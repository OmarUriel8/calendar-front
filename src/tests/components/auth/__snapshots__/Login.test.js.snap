// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Login/> Debe mostrarse correctamente 1`] = `
<div
  className="container login-container mb-5"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1 offset-md-3"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control false "
            name="email"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="correo@correo.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control false "
            name="password"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            disabled={false}
            type="submit"
            value="Login"
          />
        </div>
        <a
          href="/auth/register"
          onClick={[Function]}
        >
          ¿Aún no tienes una cuenta?, Registrate
        </a>
      </form>
    </div>
  </div>
</div>
`;
